doozerlib/dblib.py

```diff
@@ -21,44 +148,128 @@ def __init__(self, runtime, environment, dry_run=False):
         :param dry_run: If True, not data will be written
         """
         self.lock = threading.Lock()
-        self.existing_domains = {}  # maps existing domain names to True
+
         self.runtime = runtime
         self.environment = environment
         self.cmd_line = ','.join(sys.argv)
         m = hashlib.md5()
         m.update(self.cmd_line.encode('utf-8'))
         self.uniq = m.hexdigest()  # help guarantee record uniqueness
 
-        self.client = None
-        if environment and not dry_run:
-            try:
-                self.client = boto3.client('sdb', region_name='us-east-1')
-                for name in self.client.list_domains()['DomainNames']:
-                    self.existing_domains[name] = True
-            except:
-                traceback.print_exc()
-                runtime.logger.error('Unable to acquire simpledb client; please check AWS creds')
-
-    def create_domain_if_missing(self, domain):
-        if self.client:
-            with self.lock:
-                if domain in self.existing_domains:
-                    # Already exists
-                    return
-                self.client.create_domain(
-                    DomainName=domain,
-                )
-                self.existing_domains[domain] = True
+        self.connection = None
+
+        self.column_identifier = self.create_column_identifier()
+
+    @staticmethod
+    def create_column_identifier():
+        int_handler = IntegerIdentifier()
+        float_handler = FloatIdentifier()
+        datetime_handler = DateTimeIdentifier()
+        varchar_handler = VarcharIdentifier()
+
+        int_handler.set_next_handler(float_handler).set_next_handler(datetime_handler).set_next_handler(varchar_handler)
+        return int_handler
+
+    def check_table_exist(self, table_name):
+        cursor = self.connection.cursor()
+        cursor.execute("""
+            SELECT COUNT(*)
+            FROM information_schema.tables
+            WHERE table_name = '{0}'
+            """.format(table_name.replace('\'', '\'\'')))
+        if cursor.fetchone()[0] == 1:
+            cursor.close()
+            return True
+
+        cursor.close()
+        return False
+
+    def get_missing_columns(self, attr_payload, table_name):
+
+        cursor = self.connection.cursor()
+        cursor.execute(f"show columns from {table_name}")
+        existing_columns = set()
+        payload_columns = set()
+
+        for data in cursor:
+            existing_columns.add(data[0])
+
+        for column in attr_payload:
+            payload_columns.add(column)
+
+        missing_columns = list(payload_columns.difference(existing_columns))
+        return_columns = []
+        for column in missing_columns:
+            return_columns.append((column, self.column_identifier.identify(attr_payload[column])))
```

Comment:
As mentioned elsewhere, since mysql supports non-str data types, change Record.set so that it doesn't covert things to strings anymore. That should mean we don't need anything fancy to determine column type.
```python
def identify_column(val):
    pytype = type(val)
    if  pytype == 'int':
        return 'BIGINT'
    elif pytype == 'float':
        return 'FLOAT'
    elif pytype == `datetime.datetime`:
        return 'DATETIME'
    elif pytype == 'str':
        return 'VARCHAR'
    # .....etc
    else:
        raise IOError(f'Unexpect attribute type: {pytype}')       
```

Change all record set operations in the rest of the code to send in real ints, datetimes, etc.

Reply:
Agreed. But if there are two places where Record.set is called like this, Record.set("label.number", "1") first and Record.set("label.number", 2) after that and assuming label.number is a new column we would identify label.number as varchar instead of int for all the subsequent records.  There are possible flaws with my implementation as well but I guess we will have to handle some of the things manually as well since a lot of schema related things are happening dynamically which shouldn't happen ( like if the current implementation identifies a new column type as int but all the subsequent records are sending float for that value the implementation fails.)

Reply:
I consider a Record.set invocation with the wrong datatype a bug in doozer and nothing I think you should have to accommodate.

Reply:
@jupierce agreed this was a better approach. Incorporated this in the code. Now it's the responsibility of the client to send correct type during Record.set